wicketd/src/update_tracker.rs

```diff
@@ -380,31 +382,158 @@ impl UpdateDriver {
+        // long as it matches the trampoline image we're expecting the SP to be
+        // pulling, but it's possible we could be seeing stale status from a
+        // previous update attempt with the same image.
+        //
+        // To start, _clear out_ the most recent status that MGS may have
+        // cached, so we don't see any stale progress from a previous update
+        // through this SP. If somehow we've lost the race and our SP is already
+        // actively requesting host blocks, this will discard a real progress
+        // message, but that's fine - in that case we expect to see another real
+        // one imminently. It's possible (but hopefully unlikely?) that the SP
+        // is getting its phase two image from the _other_ scrimlet's MGS
+        // instance, in which case we will get no progress info at all until
+        // installinator starts reporting in.
+        //
+        // Throughout this function, we do not fail if a request to MGS fails -
+        // these are all "best effort" progress; our real failure mode is if
+        // installinator tells us it has failed.
+        if let Err(err) = self
+            .mgs_client
+            .sp_host_phase2_progress_delete(self.sp.type_, self.sp.slot)
+            .await
+        {
+            warn!(
+                self.log, "failed to clear SP host phase2 progress";
+                "err" => %err,
+            );
+        }
+
+        // TODO replace this with whatever we're doing for installinator
+        // reporting progress to us through our artifact server; for now, this
+        // function will never return. We will probably restructure this loop
+        // too (e.g., if we have a future we can await).
+        let got_first_installinator_progress = |_update_id| false;
+
+        self.set_current_update_state(
+            UpdateStateKind::WaitingForTrampolineImageDelivery {
+                artifact: plan.trampoline_phase_1.id.clone(),
+                progress: HostPhase2Progress::None,
+            },
+        );
+        while !got_first_installinator_progress(self.update_id) {
+            match self
+                .mgs_client
+                .sp_host_phase2_progress_get(self.sp.type_, self.sp.slot)
+                .await
+                .map(|response| response.into_inner())
+            {
+                Ok(HostPhase2Progress::Available {
+                    age,
+                    image_id,
+                    offset,
+                    total_size,
+                }) => {
+                    // Does this image ID match the one we uploaded? If so,
+                    // record our current progress; if not, this is probably
+                    // stale data from a past update, and we have no progress
+                    // information.
+                    let progress = if image_id == uploaded_trampoline_phase2_id
+                    {
+                        HostPhase2Progress::Available {
```

Comment:
This should be impractical if not impossible on our network, but would it be useful to only update if the most recent update is a later update? Does MGS already do this check?

Reply:
I'm not sure what you mean by "only update if the most recent update is a later update" - can you clarify? :)

In general I don't think MGS can do much checking of these progress messages; they're driven by the SP (which itself is driven by the host requesting data, in this case), and it's even possible (albeit unlikely, at least for MUP) that the SP is alternating between MGS instances if they both have the necessary image.

Reply:
Oh, sorry, I realize that comment was totally unclear. What I meant was, we could check (probably here in wicketd) if `offset` is larger than the prior saved offset, so that progress doesn't go backwards for the same install.